bundle-workflow/src/test_workflow/test_recorder/test_recorder.py

```diff
@@ -0,0 +1,123 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+import logging
+import os
+import shutil
+from typing import Dict
+
+import yaml
+
+from test_workflow.test_recorder.test_recorder_builder import \
+    TestRecorderBuilder
+
+
+class Singleton(type):
+    """Metaclass."""
+    _instances: Dict[type, object] = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        else:
+            logging.info("TestRecorder is a singleton class, retrieving existing instance")
+        return cls._instances[cls]
+
+    def get_instance(cls):
+        if cls._instances is not None:
+            return cls._instances[cls]
+
+
+class TestRecorder(object, metaclass=Singleton):
+    ACCEPTABLE_TEST_TYPES = ["integ-test", "bwc-test", "perf-test"]
+
+    def __init__(self, test_run_id, test_type, location):
+        self.test_run_id = test_run_id
+        self.test_type = test_type
+        self.location = os.path.join(location, "test-recorder")
+        logging.info(f"TestRecorder recording logs in {self.location}")
+        if self.test_type not in self.ACCEPTABLE_TEST_TYPES:
+            raise ValueError(
+                f"{self.test_type} is not invalid. Accceptable test_type are {self.ACCEPTABLE_TEST_TYPES}")
+
+    def record_cluster_logs(self, cluster_type, test_recorder_builder: TestRecorderBuilder):
+        acceptable_cluster_type = ["remote", "local"]
+        if cluster_type not in acceptable_cluster_type:
+            raise ValueError(f"{cluster_type} is not invalid. Accceptable cluster_type are {acceptable_cluster_type}")
+        base = self.__create_base_folder_structure(test_recorder_builder.component_name,
+                                                   test_recorder_builder.component_test_config)
+        folder_name = cluster_type + "-cluster-logs"
+        dest_directory = os.path.join(base, folder_name)
+        os.makedirs(dest_directory, exist_ok=False)
+        logging.info(
+            f"Recording {cluster_type} cluster logs for {test_recorder_builder.component_name} "
+            f"with {test_recorder_builder.component_test_config} config in {os.path.realpath(dest_directory)}")
+        self.__generate_std_files(test_recorder_builder.stdout, test_recorder_builder.stderr,
+                                  os.path.realpath(dest_directory))
+        if cluster_type == "local":
```

Comment:
+1

Reply:
It was separate before but clubbed it together after the conversation here: https://github.com/opensearch-project/opensearch-build/pull/467#discussion_r708717176

Reply:
[Discussed Offline] The usages pattern of TestRecorder are making this strange, I would expect that somewhere a TestCluster 
 accepts a TestRecorder in its constructor, then before the cluster vanishes it calls a method on the test recorder 'tr.saveMyUsefulLogs(logData)', this would put the responsibility of the log reporting onto the cluster which would have the most inside into where the logs are and how to get them.  This also means that one can be implemented before the other and the TestRecorder doesn't need to know how this works.

In the current flow, the TestRecorder is responsible for finding these logs which is creating based on the type passed in.

Reply:
Why would we need inner classes here then? The user would need to generate 2 different objects. Wont that complicate this simple tasks?
I think for now its better we just have 2 different methods inside TestRecorder class `record_local_cluster_logs` and `record_remote_cluster_logs` to simplify things and the user can call them when they need to save relevent logs
`tr.record_local_cluster_logs(logData)`

Reply:
Because of single responsibility principle. Mentally it's easier to understand what code does if it does 1 thing, and 1 thing well. So every time there's an `if` statement it's an opportunity to extract the responsibilities performed by the code into a class. Consider this:

```
self.record_logs()
```

vs.

```
if (cluster_type == "local"):
  self.record_local_logs()
elsif (cluster_type == "remote"):
  self.record_remote_logs()
```

The first one removes the need for the reader to think about what's going on with recording logs, but the entire purpose of what is being done is preserved.

If you just extract the `if` into a `record_logs` method, you're displacing the problem, so we move that internal logic into a `LogRecorderLocal` and `LogRecorderRemote` and can write:

```
LogRecorder.create(cluster_type).record()
```

The immediate advantage is that each log recorder type can be tested independently and that new cluster types can be added without modifying the test recorder code.